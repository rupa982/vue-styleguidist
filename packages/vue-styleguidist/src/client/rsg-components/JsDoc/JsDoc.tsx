import React from 'react'
import PropTypes from 'prop-types'
import * as Rsg from 'react-styleguidist'
import map from 'lodash/map'
import Markdown from 'rsg-components/Markdown'
import capitalize from 'lodash/capitalize'
import Styled, { JssInjectedProps } from 'rsg-components/Styled'
import { Param } from 'vue-docgen-api'

export interface TagProps {
	deprecated?: Param[]
	see?: Param[]
	link?: Param[]
	author?: Param[]
	version?: Param[]
	since?: Param[]
}

const styles = ({ space, color, fontFamily }: Rsg.Theme) => ({
	wrapper: {
		color: color.base,
		fontSize: 'inherit',
		lineHeight: 1.5
	},
	name: {
		marginRight: space[1]
	}
})

const list = (array: Param[]) => array.map(item => item.description).join(', ')
const paragraphs = (array: Param[]) => array.map(item => item.description).join('\n\n')

const fields: Record<keyof TagProps, (v: Param[]) => string> = {
	deprecated: (value: Param[]) => `${value[0].description}`,
	see: (value: Param[]) => paragraphs(value),
	link: (value: Param[]) => paragraphs(value),
	author: (value: Param[]) => list(value),
	version: (value: Param[]) => `${value[0].description}`,
	since: (value: Param[]) => `${value[0].description}`
}

export const JsDoc: React.FC<TagProps & JssInjectedProps> = ({ classes, ...props }) => {
	return (
		<>
			{map(fields, (format: (v: Param[]) => string, field: keyof TagProps) => {
				const value = props[field]
				if (!value || !Array.isArray(value)) return null
				return (
					<div className={`vsg-jsdoc-tag ${classes.wrapper}`} key={field}>
						<span className={`vsg-tag-name ${classes.name}`}>{capitalize(field)}:</span>
						<span className={`vsg-tag-value ${classes.value}`}>
							{<Markdown text={format(value) || ''} inline />}
						</span>
					</div>
				)
			})}
		</>
	)
}

JsDoc.propTypes = {
	classes: PropTypes.objectOf(PropTypes.string.isRequired).isRequired,
	deprecated: PropTypes.array,
	see: PropTypes.array,
	link: PropTypes.array,
	author: PropTypes.array,
	version: PropTypes.array,
	since: PropTypes.array
}

export default Styled<TagProps & JssInjectedProps>(styles)(JsDoc)
