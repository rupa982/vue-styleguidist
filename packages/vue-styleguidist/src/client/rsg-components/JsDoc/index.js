import React from 'react'
import PropTypes from 'prop-types'
import map from 'lodash/map'
import capitalize from 'lodash/capitalize'
import Markdown from 'rsg-components/Markdown'
import Styled from 'rsg-components/Styled'

const list = array => array.map(item => item.description).join(', ')
const paragraphs = array => array.map(item => item.description).join('\n\n')

const fields = {
	deprecated: value => value[0].description,
	see: value => paragraphs(value),
	link: value => paragraphs(value),
	author: value => list(value),
	version: value => value[0].description,
	since: value => value[0].description
}

const styles = ({ space, color, fontFamily }) => ({
	wrapper: {
		color: color.base,
		fontFamily: fontFamily.base,
		fontSize: 'inherit',
		lineHeight: 1.5
	},
	name: {
		marginRight: space[1]
	}
})

function JsDoc(props) {
	const { classes } = props
	return map(
		fields,
		(format, field) =>
			props[field] && (
				<div className={`vsg-jsdoc-tag ${classes.wrapper}`} key={field}>
					<span className={`vsg-tag-name ${classes.name}`}>{capitalize(field)}:</span>
					<span className={`vsg-tag-value ${classes.value}`}>
						{<Markdown text={format(props[field])} inline />}
					</span>
				</div>
			)
	).filter(Boolean)
}

JsDoc.propTypes = {
	classes: PropTypes.object.isRequired,
	deprecated: PropTypes.array,
	see: PropTypes.array,
	link: PropTypes.array,
	author: PropTypes.array,
	version: PropTypes.array,
	since: PropTypes.array
}

export default Styled(styles)(JsDoc)
