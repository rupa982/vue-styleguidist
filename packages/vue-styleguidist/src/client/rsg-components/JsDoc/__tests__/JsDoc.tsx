import React from 'react'
import { render } from '@testing-library/react'
import JsDoc from '../JsDoc'

const tags = {
	deprecated: [
		{
			title: 'description',
			description: 'Use *another* method'
		}
	],
	version: [
		{
			title: 'version',
			description: '2.0.0'
		}
	],
	since: [
		{
			title: 'since',
			description: '1.0.0'
		}
	],
	author: [
		{
			title: 'author',
			description: '[Author 1](#TestLink)'
		},
		{
			title: 'author',
			description: '[Author 2](#TestLink2)'
		}
	],
	see: [
		{
			title: 'see',
			description: '[See 1](#TestLink)'
		},
		{
			title: 'see',
			description: '[See 2](#TestLink2)'
		}
	],
	link: [
		{
			title: 'link',
			description: '[Link 1](#TestLink)'
		}
	]
}

describe('JsDoc', () => {
	it('should render Markdown', () => {
		const actual = render(<JsDoc {...tags} />)

		expect(actual.container).toMatchSnapshot()
	})

	it('should render null for empty tags', () => {
		const actual = render(<JsDoc />)

		expect(actual.container.innerHTML).toBe('')
	})
})
