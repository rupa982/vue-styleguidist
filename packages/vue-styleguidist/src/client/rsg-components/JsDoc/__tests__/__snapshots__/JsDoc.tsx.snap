// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JsDoc should render Markdown 1`] = `
<div>
  <div
    class="vsg-jsdoc-tag rsg--wrapper-0"
  >
    <span
      class="vsg-tag-name rsg--name-1"
    >
      Deprecated
      :
    </span>
    <span
      class="vsg-tag-value undefined"
    >
      <span
        class="rsg--text-3 rsg--inheritSize-4 rsg--baseColor-8"
      >
        Use 
        <em
          class="rsg--text-3 rsg--inheritSize-4 rsg--baseColor-8 rsg--em-10"
        >
          another
        </em>
         method
      </span>
    </span>
  </div>
  <div
    class="vsg-jsdoc-tag rsg--wrapper-0"
  >
    <span
      class="vsg-tag-name rsg--name-1"
    >
      See
      :
    </span>
    <span
      class="vsg-tag-value undefined"
    >
      <div>
        <span
          class="rsg--text-3 rsg--inheritSize-4 rsg--baseColor-8"
        >
          <a
            class="rsg--link-13"
            href="#TestLink"
          >
            See 1
          </a>
        </span>
        <span
          class="rsg--text-3 rsg--inheritSize-4 rsg--baseColor-8"
        >
          <a
            class="rsg--link-13"
            href="#TestLink2"
          >
            See 2
          </a>
        </span>
      </div>
    </span>
  </div>
  <div
    class="vsg-jsdoc-tag rsg--wrapper-0"
  >
    <span
      class="vsg-tag-name rsg--name-1"
    >
      Link
      :
    </span>
    <span
      class="vsg-tag-value undefined"
    >
      <span
        class="rsg--text-3 rsg--inheritSize-4 rsg--baseColor-8"
      >
        <a
          class="rsg--link-13"
          href="#TestLink"
        >
          Link 1
        </a>
      </span>
    </span>
  </div>
  <div
    class="vsg-jsdoc-tag rsg--wrapper-0"
  >
    <span
      class="vsg-tag-name rsg--name-1"
    >
      Author
      :
    </span>
    <span
      class="vsg-tag-value undefined"
    >
      <span
        class="rsg--text-3 rsg--inheritSize-4 rsg--baseColor-8"
      >
        <a
          class="rsg--link-13"
          href="#TestLink"
        >
          Author 1
        </a>
        , 
        <a
          class="rsg--link-13"
          href="#TestLink2"
        >
          Author 2
        </a>
      </span>
    </span>
  </div>
  <div
    class="vsg-jsdoc-tag rsg--wrapper-0"
  >
    <span
      class="vsg-tag-name rsg--name-1"
    >
      Version
      :
    </span>
    <span
      class="vsg-tag-value undefined"
    >
      <span
        class="rsg--text-3 rsg--inheritSize-4 rsg--baseColor-8"
      >
        2.0.0
      </span>
    </span>
  </div>
  <div
    class="vsg-jsdoc-tag rsg--wrapper-0"
  >
    <span
      class="vsg-tag-name rsg--name-1"
    >
      Since
      :
    </span>
    <span
      class="vsg-tag-value undefined"
    >
      <span
        class="rsg--text-3 rsg--inheritSize-4 rsg--baseColor-8"
      >
        1.0.0
      </span>
    </span>
  </div>
</div>
`;
